File: 0046-permutations/0046-permutations.go
func permute(nums []int) [][]int {
    var result [][]int
    var current []int
    used := make([]bool, len(nums))
    sort.Ints(nums)
    backtrack(nums, current , &result , used)
    return result
}

func backtrack(nums []int, current []int, result *[][]int, used[]bool){
    if len(current) == len(nums){
            temp := make([]int, len(current))
            copy(temp, current)
            *result = append(*result, temp)
            return
    }

    for i := 0  ; i <len(nums); i++{
        if used[i]{
            continue
        }
        used[i] = true
        current = append(current, nums[i])
        backtrack(nums, current , result , used)
        used[i]= false
        current = current[:len(current)-1]
    }
}

File: 0046-permutations/0046-permutations.rs
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut current = Vec::new();
        let mut result = Vec::new();
        let mut nums = nums;
        nums.sort_unstable();
        let mut used = vec![false; nums.len()];
        Self::backtrack(&nums, &mut current, &mut result, &mut used);
        result
    }
    fn backtrack(nums : &[i32], current : &mut Vec<i32>, result : &mut Vec<Vec<i32>>, used :&mut[bool]){
        if current.len() == nums.len(){
            result.push(current.clone());
            return ;
        }
        for i in 0..nums.len(){
            if used[i] {
                continue;
            }
            used[i] = true;
            current.push(nums[i]);
            Self::backtrack(nums,  current, result, used);
            used[i]= false;
            current.pop();
        }


    }
}